core/shared/src/main/scala/fs2/internal/Algebra.scala

```diff
@@ -1,24 +1,29 @@
 package fs2.internal
 
-import cats.data.NonEmptyList
 import cats.~>
-import cats.effect.Sync
+import cats.effect.{Effect, Sync}
 import cats.implicits._
 import fs2._
 
+import scala.concurrent.ExecutionContext
+
 private[fs2] sealed trait Algebra[F[_],O,R]
 
+// Shared algebra between fold/uncons
+private[fs2] sealed trait AlgebraShared[F[_], O, R] extends Algebra[F, O, R]
```

Comment:
I'd like a better name here - also, let's move this trait inside Algebra companion object.

Reply:
How about `Algebra.Effectful`? Because the only constructors which aren't subtypes of it are `Output`, `Run`, and `Uncons` which by themselves are all non-effectful (stepping through an uncons may involve evaluating some nested effectful nodes of course)

Reply:
Yes. I am still playing with Scope to have specific implementation for Uncons/Fold to resolve the interrupt. But yes lets name it effectfull